active/0003-extensible-resource-deployment-observability-improvements.md

```diff
@@ -0,0 +1,230 @@
+
+To address this issue, the deployments and deployment operations APIs must be updated to include additional information, such as:
+
+- Resource symbolic name
+- Resource type
+- Deployment provider alias
+- Deployment provider name
+- Deployment provider version
+
+This additional data would enable users to identify a deployed extensible resource and map it to the corresponding resource definition in a Bicep file or an ARM template, providing users with comprehensive insights.
+
+## Detailed design
+
+> The core of the REP. Elaborate on the design and implementation with sufficient detail for someone familiar with Bicep to grasp. Provide detailed examples to illustrate how the feature is used and its implications on user experience.
+
+### `Microsoft.Resources/deployments` API changes for PUT and GET
+
+#### Adding `deploymentProviders`
+
+The `Microsoft.Resources/deployments` API will be updated to include a `deploymentProviders` array in both the PUT and GET response bodies. This modification is intended to incorporate information about deployment providers that are involved in executing the deployment to offer a more comprehensive deployment view.
+
+> It's worth noting that we cannot use the name `providers` as the property `providers` is already assigned to return Azure Resource Provider information. While an alternative approach could involve renaming the current `providers` property to `resourceProviders` for enhanced accuracy, the solution should be avoided as it constitutes a breaking change.
+
+```diff
+{
+  ...
+  "type": "Microsoft.Resources/deployments",
+  "properties": {
+    ...
+    "providers": [
+      {
+        "namespace": "Microsoft.Storage",
+        "resourceTypes": [
+          {
+            "resourceType": "storageAccounts",
+            "locations": [
+              "eastus"
+            ]
+          }
+        ]
+      }
+    ],
++   "deploymentProviders": [
++     {
++       "alias": "az",
++       "name": "AzureResourceManager",
++       "version": "1.0.0"
++     }  ,
++     {
++       "alias": "kubernetes",
++       "name": "Kubernetes",
++       "version": "1.27.8"
++     }
+    ]
+  }
+}
+```
+
+#### Updating the `outputResources` to include more identifiable information
+
+To improve the current structure where each `outputResources` object only has a resource `id`, it's essential to add more properties for identifying extensible resources, which lack resource IDs. These additions should include the symbolic name, the resource type, and the deployment provider alias.
+
+```diff
+{
+  ...
+  "type": "Microsoft.Resources/deployments",
+  "properties": {
+    "outputResources": [
+      // Azure resource
+      {
+        "id": "/subscriptions/00000000-0000-0000-0000-000000000001/resourceGroups/my-resource-group/providers/Microsoft.Storage/storageAccounts/my-storage-account"
+      },
+      // Extensible resource
+      {
```

Comment:
AFAIK `outputResources` today captures resources from nested deployments as well. This works because the `resourceId` is globally unique within a tenant. The same is not true for symbolic names - do we need something to identify which nested deployment they relate to?

Reply:
Good catch, I missed that part. It seems we would need to include the deployment name as well.

Reply:
We may also need to add deployment name to `deploymentProviders` to disambiguate 🤔.

Reply:
Added a `deploymentId` property to resolve this.

Reply:
This may be too much of a breaking change to justify, but another option that might be worth discussing is having the `outputResources` just return resources deployed directly, and to require the caller (i.e. Stacks) to traverse the deployment graph.

Re-opening to discuss today.

Reply:
Missed this in today's meeting, but I guess we can discuss it asynchronously here. Restricting `outputResources` to only include resources deployed in the current deployment can make the API more concise and relevant to the specific deployment context. Do we know when `outputResources` is introduced? If it's a longstanding property not originally designed for Deployment Stacks, making the breaking change is a worrying, since removing elements from an established array is a kind of behavioral change that might not be immediately apparent to users.